### PR TITLE
Adding an unsat core reduce function

### PR DESCRIPTION
**Pull request by zhanghongce**
This is to iterate through the unsat assumptions. This is used in SyGuS PDR and it guarantees to be minimal (but not minimum)

### CODE REVIEW

**Comment by zhanghongce** [comment]
I'm still working on this. Especially, there are questions about the expected behavior of duplicate assumptions.

**Review by makaimann — APPROVED** [review]
LGTM, thanks for the PR!

**Comment by makaimann** [comment]
Marking `do not merge` until you (@zhanghongce) say you're done working on it. If you're worried about duplicate assumptions, you could consider de-duplicating with an `UnorderedTermSet` first.

**Comment by zhanghongce** [comment]
Thanks! I think it is now ready for review.

File: src/utils.cpp
Comment (makaimann) [inline src/utils.cpp:295]: minor: can delete this line

File: src/utils.cpp
Comment (makaimann) [inline src/utils.cpp:319]: minor: can remove this line

File: src/utils.cpp
Comment (makaimann) [inline src/utils.cpp:329]: minor: can remove this line

File: src/utils.cpp
Comment (makaimann) [inline src/utils.cpp:290]: do you actually need to populate the set here before the loop?

File: tests/test-unsat-core-reducer.cpp
Comment (makaimann) [inline tests/test-unsat-core-reducer.cpp:74]: This isn't too important, but it's more idiomatic to use:
`EXPECT_EQ(<arg1>, <arg2>)` and
`EXPECT_NE(<arg1>, <arg2>)`

**Review by makaimann — APPROVED** [review]
Looks good, thanks @zhanghongce! I just left a few minor comments that you can address or not.

File: src/utils.cpp
Comment (zhanghongce) [reply src/utils.cpp:290]: This was for the case that not a single assumption can be removed. (In that case no UNSAT core extraction is executed)
But maybe I can write in another way that won't rely on the final `core_set`... let me change to that ...

**Comment by zhanghongce** [comment]
The commit 013f199 will resolve #151

**Comment by makaimann** [comment]
> The commit [013f199](https://github.com/makaimann/smt-switch/commit/013f199fde65fb8a756af4ec1a31b731284db57e) will resolve #151

Good catch! That's definitely a problem.

**Comment by makaimann** [comment]
@ahmed-irfan do you want to make a pass when you have a chance? Looks good to me, but a second pair of eyes is always good.

File: include/utils.h
Comment (ahmed-irfan) [inline include/utils.h:135]: add comment about the result

File: src/utils.cpp
Comment (ahmed-irfan) [inline src/utils.cpp:313]: just a style thing: add brackets here

File: src/utils.cpp
Comment (ahmed-irfan) [inline src/utils.cpp:315]: ditto

File: src/utils.cpp
Comment (ahmed-irfan) [inline src/utils.cpp:311]:
```suggestion
        while (bool_assump_pos != bool_assump.end()) {
```

File: src/utils.cpp
Comment (ahmed-irfan) [inline src/utils.cpp:304]: where do we remove this assumption? `assump_pos`

File: src/utils.cpp
Comment (ahmed-irfan) [inline src/utils.cpp:260]: maybe rename this to local_assump

**Review by ahmed-irfan — CHANGES_REQUESTED** [review]
Left some comments

File: src/utils.cpp
Comment (makaimann) [reply src/utils.cpp:304]: So I think it will be removed below because it by definition is not in the unsat core (since it wasn't assumed in this iteration). Although, one minor thing to note is that `vector::erase` is not very efficient. It might even be better to clear the vector and just add things back into it.

Comment (ahmed-irfan) [reply src/utils.cpp:304]: Right.

**Review by ahmed-irfan — APPROVED** [review]
LGTM. Thank for adding this

**Pull request merged**
